### ISSUE TITLE
make_mesh_3 returns empty mesh for my domain with planar 1D features

### ISSUE THREAD

**Issue by jpanetta** [CGAL/cgal#1501]
I'm using make_mesh_3 on a domain defined by a smooth signed distance function clipped to a cube. When I add add planar feature curves at the surface-cube intersection, make_mesh_3 often returns a complex with no cells.

I attached the simplest example I could get to fail (I couldn't reproduce the problem with a single clipped sphere or an assembly of axis-aligned cylinders). The example consists of a smooth union of off-axis cylinders. I hard-coded feature curves for the clipping cube intersection that I extracted with marching squares.

By perturbing the amount of smoothing (slightly changing the joint geometry) I was able to extract feature curves that meshed properly. The attached program can be run with the failing "original" feature curves, the working "perturbed" feature curves, or no feature curves:

$ ./self_contained original
got 0 tetrahedra
$ ./self_contained perturbed
got 106736 tetrahedra
$ ./self_contained none
got 149256 tetrahedra

There is more information on my [cgal-discuss post](http://cgal-discuss.949826.n4.nabble.com/Bug-in-make-mesh-3-on-domain-with-planar-1D-features-Returning-empty-mesh-td4662300.html). Note that the code attached here differs from my cgal-discuss attachment: it fixes a mistake where my domain bounding sphere's center wasn't inside the geometry (though this didn't affect the issue).

[failing_example.zip](https://github.com/CGAL/cgal/files/496220/failing_example.zip)

**Comment by lrineau** [comment]
Thanks for the detailed report. I am working on it now.

**Comment by lrineau** [comment]
I found out what happens. That was pretty surprising!

For the initialization of the triangulation, the code is in the file `CGAL-4.9/include/CGAL/make_mesh_3.h`, and the code is:
```C++
// Partial specialization of C3t3_initializer
// Handles cases where MeshDomain::Has_features is a valid type and is defined
// to CGAL::Tag_true
template < typename C3T3, typename MD, typename MC >
struct C3t3_initializer < C3T3, MD, MC, true, CGAL::Tag_true >
{
  void operator()(C3T3& c3t3,
                  const MD& domain,
                  const MC& criteria,
                  bool with_features,
                  const int nb_initial_points = -1)
  {
    if ( with_features ) {
      init_c3t3_with_features(c3t3,domain,criteria);

      // If c3t3 initialization is not sufficient (may happen if there is only
      // a planar curve as feature for example), add some surface points
      if ( c3t3.triangulation().dimension() != 3 ) {
        init_c3t3(c3t3, domain, criteria, nb_initial_points);
      }
    }
    else { init_c3t3(c3t3,domain,criteria,nb_initial_points); }
  }
};
```

The function `init_c3t3_with_features` only create initial points on the feature curves, and the function `init_c3t3`, for an implicit domain, shots randoms rays from the center of the sphere, to find out input points.

In your case, the sampling of initial points on the feature curves created a set of 926 vertices. But then the circumcenters of all the tetrahedra of the Delaunay triangulation of those vertices were all in the exterior of the meshing domain (see the black dots in figure below).

![password1](https://cloud.githubusercontent.com/assets/5746675/18914247/8ebb5106-858c-11e6-9e65-7178ce401162.png)

I thought it was theoretically possible that the set of input points was in such a way that the initial mesh was empty, but that is the first time I can see that effect on a non-trivial data set.

The solution I will commit in CGAL is the following:
```patch
diff --git a/Mesh_3/include/CGAL/make_mesh_3.h b/Mesh_3/include/CGAL/make_mesh_3.h
index 1822508..fd7e329 100644
--- a/Mesh_3/include/CGAL/make_mesh_3.h
+++ b/Mesh_3/include/CGAL/make_mesh_3.h
@@ -262,9 +262,7 @@ struct C3t3_initializer < C3T3, MD, MC, true, CGAL::Tag_true >
 
       // If c3t3 initialization is not sufficient (may happen if there is only
       // a planar curve as feature for example), add some surface points
-      if ( c3t3.triangulation().dimension() != 3 ) {
-        init_c3t3(c3t3, domain, criteria, nb_initial_points);
-      }
+      init_c3t3(c3t3, domain, criteria, nb_initial_points);
     }
     else { init_c3t3(c3t3,domain,criteria,nb_initial_points); }
   }
```

Even if the dimension of the initial triangulation is 3, then the random rays shooting will always be used, to avoid such a degenerated configuration (let's assume that the random rays shooting will not itself create a degenerated configuration).

**Comment by jpanetta** [comment]
Great, thanks for fixing this so quickly and explaining what's happening--it may be a weird corner case, but it's happening all the time in my application!

**Comment by lrineau** [comment]
I have modified your files, and pushed them to the [Gist lrineau/0b27e5024d368564ef834b5abfcc7add] (https://gist.github.com/lrineau/0b27e5024d368564ef834b5abfcc7add).

Cc @janetournois

**Comment by lrineau** [comment]
The fix from #1506 is now merged into `master` and in the release branch for CGAL-4.9.1.
